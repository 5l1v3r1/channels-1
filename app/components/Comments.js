// @flow
import React, { Component } from 'react';
import CommentForm from './CommentForm';
import CommentList from './CommentList';
import ErrorState from './core/ErrorState';
import InfiniteScroll from './core/InfiniteScroll';
import Loader from './core/Loader';
import type { PromiseAction } from '../constants/types';
import CommentType from '../dataTypes/commentType';
import styles from './Comments.scss';

type Props = {
  comments: CommentType[],
  pageToken: string,
  isFetching: boolean,
  error: boolean,
  activeVideoId: string,
  fetchComments: () => PromiseAction,
  postComment: (description: string) => PromiseAction
};

type State = {
  description: string,
  isPosting: boolean
};

export default class Comments extends Component<Props, State> {
  state = {
    description: '',
    isPosting: false
  };

  componentDidMount() {
    this.props.fetchComments();
  }

  componentWillReceiveProps(nextProps: Props) {
    const { activeVideoId: id } = this.props;
    const { activeVideoId: nextId } = nextProps;
    if (id !== nextId) {
      this.props.fetchComments();
    }
  }

  handleChange = (event: SyntheticInputEvent<HTMLTextAreaElement>): void => {
    this.setState({ description: event.currentTarget.value });
  }

  handleSubmit = (event: SyntheticInputEvent<HTMLFormElement>): void => {
    event.preventDefault();
    this.setState({ isPosting: true });
    this.props.postComment(this.state.description) // eslint-disable-line promise/catch-or-return
      .then(
        () => this.handleReset(),
        () => undefined
      )
      .then(() => this.setState({ isPosting: false }));
  }

  handleReset = (): void => {
    this.setState({ description: '' });
  }

  render() {
    const {
      comments,
      pageToken,
      isFetching,
      error,
      fetchComments
    } = this.props;
    const {
      description,
      isPosting
    } = this.state;
    if ((isFetching && !pageToken) || error) {
      return (
        <div className={styles.comments}>
          {isFetching ? (
            <Loader />
          ) : (
            <ErrorState
              color="white"
              message="Error requesting comments."
            />
          )}
        </div>
      );
    }

    return (
      <div className={styles.comments}>
        <InfiniteScroll
          pageToken={pageToken}
          isFetching={isFetching}
          loadMore={fetchComments}
        >
          <CommentForm
            description={description}
            isPosting={isPosting}
            handleChange={this.handleChange}
            handleSubmit={this.handleSubmit}
            handleReset={this.handleReset}
          />
          <CommentList
            comments={comments}
          />
        </InfiniteScroll>
      </div>
    );
  }
}
