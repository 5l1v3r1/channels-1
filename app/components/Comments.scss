@import '../app.global.scss';

.comments {
  position: relative;
  display: flex;
  flex-direction: column;
  flex-grow: 1;
  background-color: $black;
}
