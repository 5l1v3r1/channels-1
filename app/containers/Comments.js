import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as commentsActions from '../actions/comments';
import Comments from '../components/Comments';
import { getComments } from '../reducers/index';

function mapStateToProps(state, ownProps) {
  return {
    comments: getComments(state),
    pageToken: state.comments.pageToken,
    isFetching: state.comments.isFetching,
    error: state.comments.error,
    activeVideoId: ownProps.activeVideoId
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators(commentsActions, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(Comments);
