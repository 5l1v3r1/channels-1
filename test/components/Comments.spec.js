import React from 'react';
import { shallow } from 'enzyme';
import Comments from '../../app/components/Comments';

function setup() {
  const props = {
    activeVideo: {},
    comments: [],
    isFetching: false,
    error: false,
    pageToken: '',
    fetchComments: jest.fn(),
    postComment: jest.fn()
  };

  const enzymeWrapper = shallow(<Comments {...props} />);

  return {
    props,
    enzymeWrapper
  };
}

describe('Comments', () => {
  it('should call fetchComments on activeVideo id change', () => {
    const { props, enzymeWrapper } = setup();

    expect(props.fetchComments).toHaveBeenCalledTimes(0);

    enzymeWrapper.setProps({
      activeVideo: {
        id: 'XsFQEUP1MxI'
      }
    });

    expect(props.fetchComments).toHaveBeenCalledTimes(1);

    enzymeWrapper.setProps({
      activeVideo: {
        id: 'Eu35xM76kKY'
      }
    });

    expect(props.fetchComments).toHaveBeenCalledTimes(2);

    enzymeWrapper.setProps({
      activeVideo: {
        id: 'Eu35xM76kKY'
      }
    });

    expect(props.fetchComments).toHaveBeenCalledTimes(2);
  });

  it('should handle change', () => {
    const { enzymeWrapper } = setup();
    const mockEvent = {
      target: {
        value: 'Great video!'
      }
    };

    expect(enzymeWrapper.state('description')).toEqual('');

    enzymeWrapper.instance().handleChange(mockEvent);

    expect(enzymeWrapper.state('description')).toEqual('Great video!');
  });

  it('should handle reset', () => {
    const { enzymeWrapper } = setup();

    enzymeWrapper.setState({ description: 'Great video!' });

    expect(enzymeWrapper.state('description')).toEqual('Great video!');

    enzymeWrapper.instance().handleReset();

    expect(enzymeWrapper.state('description')).toEqual('');
  });
});
