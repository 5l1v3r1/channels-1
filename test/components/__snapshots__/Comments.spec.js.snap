// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comments should render self and child components 1`] = `
<div
  className="comments"
>
  <InfiniteScroll
    isFetching={false}
    loadMore={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
      }
    }
    maxHeight=""
    pageToken=""
  >
    <CommentForm
      description=""
      handleChange={[Function]}
      handleReset={[Function]}
      handleSubmit={[Function]}
      isPosting={false}
    />
    <CommentList
      comments={Array []}
    />
  </InfiniteScroll>
</div>
`;

exports[`Comments should render self and child components 2`] = `
<div
  className="comments"
>
  <InfiniteScroll
    isFetching={true}
    loadMore={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
      }
    }
    maxHeight=""
    pageToken="token"
  >
    <CommentForm
      description=""
      handleChange={[Function]}
      handleReset={[Function]}
      handleSubmit={[Function]}
      isPosting={false}
    />
    <CommentList
      comments={Array []}
    />
  </InfiniteScroll>
</div>
`;

exports[`Comments should render self with error 1`] = `
<div
  className="comments"
>
  <ErrorState
    color="white"
    message="Error requesting comments."
  />
</div>
`;

exports[`Comments should render self with loader 1`] = `
<div
  className="comments"
>
  <Loader />
</div>
`;
